src/bookshelf/src/bksw_sec.erl

```diff
@@ -21,123 +22,405 @@
+% should return status code 403 vs. 500... or at least that's what various pre-existing tests want, anyway. The
+% throw is really just a way to bail on further processing and jump to a location in the code which returns a status 403.
+% Near as I can tell from the docs, that is what the purpose of throws are in Erlang.  Although
+% they are frequently associated with error handling, docs say they are for 'nonlocal returns'
+% or 'control flow' (or at least can be used that way).
+%
+% From: https://docs.oracle.com/cd/E88353_01/html/E37845/erlang-3erl.html
+%   throw(Any) -> no_return()
+%    A non-local return from a function. If evaluated within a catch,
+%    catch returns value Any. Example:
+%
+%    > catch throw({hello, there}).
+%    {hello,there}
+%
+% From: http://www.kuaidingyue.com/ebook/erl/errors-and-exceptions.html
+%   In comparison with exits and errors, they don't really carry any 'crash that process!' intent behind them,
+%   but rather control flow.
+%
+% Anyway, we could accomplish the same thing with a complicated nest or staircase of case statements, but to
+% me a simple throw is better. The catch which returns status code 403 is at the bottom of the auth
+% function.
+%
+% Having said all of that, if throws are undesirable, I'm totally open to eliminating them if we can figure out
+% how to do it.
+%
+% Perhaps one option is to rewrite the spec and say that 500s are ok.  This changes
+% specs and forces rewriting of tests, but avoids throws.
+%
+% Perhaps another option is to set variables to 'sentinel' values on error conditions,
+% which would allow the flow to continue, and hopefully the failures could be caught
+% later on in processing. An example might be setting Host (in auth/10 directly below
+% these comments) to "" if get_req_header is undefined, and letting things continue.
+% This might make things harder to debug, vs. an early bail with a throw (or who knows,
+% maybe early bails with throws are harder to debug?), but it avoids
+% throws if that is desirable.
+%
+% Or heck, maybe structuring case statements to eliminate throws wouldn't be as bad
+% as I am assuming. Or maybe I'm missing something.
+%
+% Anyway, I'm flexible.  For now, assuming throws until we get things sorted out.
+
+auth(RequestId, Req0, #context{reqid = ReqId} = Context, Credential, XAmzDate, SignedHeaderKeysString, IncomingSignature, XAmzExpiresString, Headers0, VerificationType) ->
     try
-        mini_s3:make_signed_url_authorization(SecretKey,
-                                              erlang:list_to_existing_atom(Method),
-                                              Path,
-                                              Expires,
-                                              Headers)
+        % CODE REVIEW: Host is used in the generation of Config which is used in both verification types.
+
+        case Host = wrq:get_req_header("Host", Req0) of
+            undefined -> throw({RequestId, Req0, Context});
```

Comment:
Could we send a {halt, 403} here instead of throwing? Somethinig liike in the pattern at https://github.com/chef/chef-server/blob/master/src/bookshelf/src/bksw_sec.erl#L153-L157

Reply:
If we could simply return from the function right there - e.g. a `return` statement - we could cleanly do what you want.  But erlang doesn't have a `return`, it has throws, so I use that instead.  @stevendanna was working on a refactor to eliminate the throws, so lets see what he comes up with.

Reply:
Looking at these checks closely they seem to be verifying the header values. Webmachine lets us do header value evaluations before reaching the is_authorized method. 

The Malformed? in https://raw.githubusercontent.com/wiki/webmachine/webmachine/images/http-headers-status-v3.png. (translates into malformed_request resource call) can be a good place to add this check. That function can be added after init and before is_authorized at https://github.com/chef/chef-server/blob/master/src/bookshelf/src/bksw_wm_base.erl#L38. 

An example for ref can be: https://github.com/chef/chef-server/blob/master/src/oc_erchef/apps/oc_chef_wm/src/oc_chef_wm_base.erl#L852 (the uses different catch based on the type of error to help with the debugging).

This might help clean this method to do only presigned and http based auth and make it easier to write tests for the auth function as needed.

Happy to talk through this when you have a few.

Reply:
RE: throw/catch pattern @markan pointed some prior art at https://github.com/chef/chef-server/blob/master/src/oc_erchef/apps/oc_chef_wm/src/oc_chef_associations.erl#L199 which uses chaining in the function calls to call next function only if successful or throw the appropriately worded error (or log msg). That approach could be used in the malformed_request function.

Reply:
Re the throws, I'm thinking that it might be good to do my refactors as a potential follow up.

Reply:
I completed a basic refactor of header checks into a `malformed_request` function inside `bksw_wm_base.erl`.

Reply:
I now completed a much more extensive refactor of stuff into `malformed_request`.
https://buildkite.com/chef/chef-chef-server-master-verify/builds/2614
https://buildkite.com/chef/chef-chef-server-master-omnibus-adhoc/builds/1710#ff15d5fe-526d-428a-99d2-89a009f017a0
https://buildkite.com/chef/chef-chef-server-master-integration-test/builds/386#44bcaf15-35bb-4a8a-ba29-d7a5f804feb6